counter/src/main/java/org/infinispan/counter/impl/manager/EmbeddedCounterManager.java

```diff
@@ -87,57 +81,77 @@ private static <T> T validateCounter(Class<T> tClass, Object retVal) {
    )
    @Override
    public void remove(String counterName) {
-      removeCounter(counterName, true);
-   }
-
-   private void removeCounter(String counterName, boolean keepConfig) {
-      CounterConfiguration configuration = getConfiguration(counterName);
-      if (configuration == null) {
-         //counter not defined (cluster-wide). do nothing :)
-         return;
+      awaitCounterOperation(removeAsync(counterName, true));
+   }
+
+   public CompletionStage<Void> removeAsync(String counterName, boolean keepConfig) {
+      CompletionStage<Void> removeStage = getConfigurationAsync(counterName)
+            .thenCompose(config -> {
+               // check if the counter is defined
+               if (config == null) {
+                  return completedNull();
+               }
+
+               CompletionStage<InternalCounterAdmin> existingCounter = counters.remove(counterName);
+               // if counter instance exists, invoke destroy()
+               if (existingCounter != null) {
+                  return existingCounter.thenCompose(InternalCounterAdmin::destroy);
+               }
+
+               if (config.type() == CounterType.WEAK) {
```

Comment:
Should this only happen if the config was defined?

Reply:
correct, `config == null` is checked on line 91 before reaching this line